### PR TITLE
Sort failures by failure's line and character

### PR DESCRIPTION
**Pull request by eggggger**
#### PR checklist

- [x] Addresses an existing issue: [#3335](https://github.com/palantir/tslint/issues/3335)
- [x] New feature, bugfix, or enhancement
  - [ ] Includes tests
- [ ] Documentation update

#### CHANGELOG.md entry:

[enhancement] `Sort failures by line and character for formatters `

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @eggggger! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: src/linter.ts [hunk 1]
```diff
@@ -131,19 +131,40 @@ class Linter {
             throw new Error(`formatter '${formatterName}' not found`);
         }
 
-        const output = formatter.format(this.failures, this.fixes);
+        // Sort failures by failure's line and character.
+        const failures = this.sortFailures(this.failures);
+        const fixes = this.sortFailures(this.fixes);
```
Comment (ajafff) [inline src/linter.ts]: the fixes are actually never really used by a formatter, so they don't need to be sorted.

File: src/linter.ts [hunk 2]
```diff
@@ -131,19 +131,40 @@ class Linter {
             throw new Error(`formatter '${formatterName}' not found`);
         }
 
-        const output = formatter.format(this.failures, this.fixes);
+        // Sort failures by failure's line and character.
+        const failures = this.sortFailures(this.failures);
```
Comment (ajafff) [inline src/linter.ts]: I'd suggest to move the sorting of failures to each formatter that needs it. Formatters for machine consumption shouldn't rely on any order.
That means there are 4 formatters affected: `prose`, `verbose`, `stylish` and `code-frame`.

File: src/linter.ts [hunk 3]
```diff
@@ -131,19 +131,40 @@ class Linter {
...
+
+        const fileFailuresMap = createMultiMap(failures, (failure) => [failure.getFileName(), failure]);
+        fileFailuresMap.forEach((fileFailures) => {
+            sortedFailures = sortedFailures.concat(fileFailures.sort((failureA, failureB) => {
+                const failureALineAndCharacter = failureA.getStartPosition().getLineAndCharacter();
+                const failureBLineAndCharacter = failureB.getStartPosition().getLineAndCharacter();
+                if (failureALineAndCharacter.line === failureBLineAndCharacter.line) {
+                    return failureALineAndCharacter.character - failureBLineAndCharacter.character;
+                }
+                return failureALineAndCharacter.line - failureBLineAndCharacter.line;
```
Comment (ajafff) [inline src/linter.ts]: instead of comparing line and character, you can simply compare the position

File: src/linter.ts [hunk 4]
```diff
@@ -131,19 +131,40 @@ class Linter {
...
+            }));
+        });
+
+        return sortedFailures;
+    }
```
Comment (ajafff) [inline src/linter.ts]: As described above, I'd like to move this to the formatters. To make the compare logic reusable, consider adding it as static method to `class RuleFailure` in `src/language/rule/rule.ts`:

```ts
    public static compare(a: RuleFailure, b: RuleFailure): number {
        if (a.fileName !== b.fileName) {
            return a.fileName < b.fileName ? -1 : 1;
        }
        return a.startPosition.getPosition() - b.startPosition.getPosition();
    }
```

File: src/formatters/codeFrameFormatter.ts [hunk 5]
```diff
@@ -97,4 +98,8 @@ export class Formatter extends AbstractFormatter {
 
         return `${outputLines.join("\n")}\n`;
     }
+
+    public sortFailures(failures: RuleFailure[]): RuleFailure[] {
```
Comment (ajafff) [inline src/formatters/codeFrameFormatter.ts]: these methods don't need to be public

File: src/formatters/codeFrameFormatter.ts [hunk 6]
```diff
@@ -97,4 +98,8 @@ export class Formatter extends AbstractFormatter {
...
+        return failures.slice().sort(RuleFailure.compare);
```
Comment (ajafff) [inline src/formatters/codeFrameFormatter.ts]: Personally, I would not write a method for this simple statement. If you want to keep the method, it makes sense to move it to `AbstractFormatter` as you already noted. Works for me either way.

**Comment by ajafff** [comment]
@eggggger You're almost done. CI fails due to linting errors:

```
/home/ubuntu/tslint/src/formatters/stylishFormatter.ts
ERROR: 42:47  semicolon        Missing semicolon

/home/ubuntu/tslint/src/language/rule/rule.ts
ERROR: 261:5  member-ordering  Declaration of public static method not allowed after declaration of public instance method. Instead, this should come after private instance fields.
```

you can also run linting locally with `yarn lint`

**Comment by ajafff** [comment]
Thanks @eggggger

**Pull request merged**